pkg/collections/offledger/dcas/client/dcasclient.go

```diff
@@ -6,67 +6,326 @@ SPDX-License-Identifier: Apache-2.0
+	"github.com/trustbloc/fabric-peer-ext/pkg/collections/offledger/dcas/ipfsdatastore"
+	"github.com/trustbloc/fabric-peer-ext/pkg/thirdparty/ipfs/go-ipfs/core/coredag"
 )
 
+const (
+	balancedLayout = "balanced"
+	trickleLayout  = "trickle"
+)
+
+type config interface {
+	GetDCASMaxLinksPerBlock() int
+	IsDCASRawLeaves() bool
+	GetDCASMaxBlockSize() int64
+	GetDCASBlockLayout() string
+}
+
+type layoutStrategy = func(db *unixfshelpers.DagBuilderHelper) (ipld.Node, error)
+
 // DCASClient allows you to put and get DCASClient from outside of a chaincode
 type DCASClient struct {
-	client   *olclient.Client
-	ns, coll string
+	config
+	dagService format.DAGService
+	layout     layoutStrategy
+}
+
+func createOLWrappedClient(cfg config, channelID, ns, coll string, providers *olclient.ChannelProviders) (*DCASClient, error) {
+	return createClient(cfg, ipfsdatastore.NewOLClientWrapper(ns, coll, olclient.New(channelID, providers)))
 }
 
-// New returns a new DCAS client
-func New(channelID, ns, coll string, providers *olclient.ChannelProviders) *DCASClient {
+func createCCStubWrappedClient(cfg config, coll string, stub shim.ChaincodeStubInterface) (*DCASClient, error) {
+	return createClient(cfg, ipfsdatastore.NewStubWrapper(coll, stub))
+}
+
+func createClient(cfg config, ds datastore.Batching) (*DCASClient, error) {
+	layout, err := resolveLayoutStrategy(cfg)
+	if err != nil {
+		return nil, err
+	}
+
 	return &DCASClient{
-		ns:     ns,
-		coll:   coll,
-		client: olclient.New(channelID, providers),
+		config: cfg,
+		layout: layout,
+		dagService: dag.NewDAGService(
+			blockservice.NewWriteThrough(
+				blockstore.NewBlockstore(ds), nil,
+			),
+		),
+	}, nil
+}
+
+// Put stores the given content and returns the content ID (CID) for the value
+func (d *DCASClient) Put(data io.Reader, opts ...Option) (string, error) {
+	options := getOptions(opts)
+	switch options.nodeType {
+	case FileNodeType:
+		return d.putFile(data, options)
+	case ObjectNodeType:
+		fallthrough
+	case "":
+		return d.putObject(context.Background(), data, options)
+	default:
+		return "", fmt.Errorf("unsupported node type [%s]", options.nodeType)
+	}
+}
+
+// Delete deletes the values for the given content IDs.
+func (d *DCASClient) Delete(ids ...string) error {
+	cids := make([]cid.Cid, len(ids))
+	for i, key := range ids {
+		cID, err := cid.Decode(key)
+		if err != nil {
+			return err
+		}
+
+		cids[i] = cID
 	}
+
+	return d.dagService.RemoveMany(context.Background(), cids)
 }
 
-// Put puts the data and returns the content ID (CID) for the value
-func (d *DCASClient) Put(data io.Reader) (string, error) {
-	bytes, err := ioutil.ReadAll(data)
+// Get retrieves the value for the given content ID.
+func (d *DCASClient) Get(id string, w io.Writer) error {
+	ctx := context.Background()
+
+	nd, err := d.getNode(ctx, id)
 	if err != nil {
-		return "", err
+		logger.Infof("Error getting node for CID [%s]: %s", id, err)
+
+		return err
+	}
+
+	if nd == nil {
```

Comment:
Should this be not found error?

Reply:
I'm trying to keep it backward compatible. It currently returns nil if the value is not found.